docs: jwt signing keys (#37027)
supabase/supabase: apps/docs/content/guides/auth/signing-keys.mdx

> ---

Reviewer: remember to add this to the nav

id: 'auth-signing-keys'
title: 'JWT Signing Keys'
subtitle: 'Best practices on managing keys used by Supabase Auth to create and verify JSON Web Tokens'
---

Supabase Auth continuously issues a new JWT for each user session, for as long as the user remains signed in. Check the comprehensive guide on [Sessions](/docs/guides/sessions) for all the details about how this works. Read our guide on [JWTs](/docs/guides/jwts) if you are not familiar with the basics before continuing.

## Overview

> When a JWT is issued by Supabase Auth, the key used to create its [signature](https://en.wikipedia.org/wiki/Digital_signature) is known as the signing key. Supabase provides has two systems for dealing with signing keys: the Legacy system based on the JWT secret, and the new Signing keys system.

Reviewer: ```suggestion
When a JWT is issued by Supabase Auth, the key used to create its [signature](https://en.wikipedia.org/wiki/Digital_signature) is known as the signing key. Supabase provides two systems for dealing with signing keys: the Legacy system based on the JWT secret, and the new Signing keys system.
```

| Zero-downtime rotation | Downtime, sometimes being significant. | No downtime, as each rotation step is independent and reversible. |
| Users signed out during rotation | Currently active users get immediately signed out. | No users get signed out. |
| Independence from API keys | `anon` and `service_role` must be rotated simultaneously. | [Publishable and secret API keys](/docs/guides/api/api-keys) no longer are based on the JWT signing key. |
| Security compliance frameworks (SOC2, etc.) | Difficult to remain aligned as the secret can be extracted from Supabase. | Easier alignment as the private key or shared secret can't be extracted. |

## Getting started

You can easily start migrating away from the legacy JWT secret through the Supabase dashboard. This process does not cause downtime for your application.

1. Start off by clicking the Migrate JWT secret button on the [JWT signing keys](/dashboard/project/_/settings/jwt/signing-keys) page. This step will import the existing legacy JWT secret into the new JWT signing keys system. Once this process completes, you will no longer be able to rotate the legacy JWT secret using the old system.

> 2. Simultaneously, we're creating a new asymmetric JWT signing key for you to rotate into. This key starts off as standby key -- meaning it's being advertised as a key that Supabase Auth will use in the future to create JWTs.

Reviewer: ```suggestion
2. Simultaneously, we're creating a new asymmetric JWT signing key for you to rotate to. This key starts off as standby key -- meaning it's being advertised as a key that Supabase Auth will use in the future to create JWTs.
```

> 3. If you're not ready to fully move to JWT signing keys at this point, you can stop here without any issue. If you wish to use a different signing key -- either to use a diffrent signing algorithm (RSA, Elliptic Curve or shared secret) or to import a private key or shared secret you already have -- feel free to move the standby key to *Previously used* before finally moving it to *Revoked.*

Reviewer: ```suggestion
3. If you're not ready to fully move to JWT signing keys at this point, you can stop here without any issue. If you wish to use a different signing key -- either to use a different signing algorithm (RSA, Elliptic Curve or shared secret) or to import a private key or shared secret you already have -- feel free to move the standby key to *Previously used* before finally moving it to *Revoked.*
```

...

## Frequently asked questions

### Why is it not possible to extract the private key or shared secret from Supabase?

> You can only extract the legacy JWT secret. Once you've moved to using the JWT signing keys feature extracting of the private key or shared secret from Supabase is not possible. This ensures that noone in your organization is able to impersonate your users or gain privileged access to your project's data.

Reviewer: ```suggestion
You can only extract the legacy JWT secret. Once you've moved to using the JWT signing keys feature extracting of the private key or shared secret from Supabase is not possible. This ensures that no one in your organization is able to impersonate your users or gain privileged access to your project's data.
```

This ensures that your project maintains high alignment with security compliance frameworks like SOC2, PCI-DSS, ISO27000, HIPAA and others.

### How do I make my own JWTs if I'm not able to extract the private key or shared secret?

If you wish to make your own JWTs or have access to the private key or shared secret used by Supabase, you can create a new JWT signing key by importing a private key or setting a shared secret yourself.

Note that this will still not be extractable from Supabase, so make sure you store it in a secure place.

### Why do I have to wait 5 minutes when moving the key through different states?

> Changing a JWT signing key's status sets of many changes inside the Supabase platform. To ensure a consistent setup, most actions that change the state of a JWT signing key are throttled for approximately 5 minutes.

Reviewer: ```suggestion
Changing a JWT signing key's status sets off many changes inside the Supabase platform. To ensure a consistent setup, most actions that change the state of a JWT signing key are throttled for approximately 5 minutes.
```
